File: thrift/scala/src/main/scala/com/colobu/rpctest/AppClient.scala
package com.colobu.rpctest

import java.util.concurrent.{CountDownLatch, Executors}

import org.apache.thrift.protocol.TBinaryProtocol
import org.apache.thrift.transport.TSocket

object AppClient extends App{
  try {
    val transport = new Array[TSocket](20)
    val client = new Array[Greeter.Client](20)
    //new Greeter.Client(protocol)

    //warm up
    val name = "world"

    for (i <- 0 to 19) {
      transport(i) = new TSocket("localhost", 9090)
      transport(i).open()

      val protocol = new TBinaryProtocol(transport(i))
      client(i) = new Greeter.Client(protocol)
      client(i).sayHello(name)
    }

    var sync = true
    if (args.length > 0) {
      sync = args(0).toBoolean
    }

    if (sync) {
      syncTest(client(0), name)
    }
    else {
      asyncTest(client, name)
    }

    for (i <- 0 to 19) {
      transport(i).close()
    }
  }
  catch {
    case e: Throwable => e.printStackTrace()
  }


  def syncTest(client:Greeter.Client , name:String):Unit = {
    val t = System.nanoTime()
    for (i <- 1 to 10000) {
      client.sayHello(name)
    }
    println("took: " + (System.nanoTime() - t) / 1000000 + " ms")
  }

  def asyncTest(client:Array[Greeter.Client] , name:String):Unit = {
    val latch = new CountDownLatch(10000)
    val pool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2)
    val t = System.nanoTime()
    for (i <- 1 to 10000) {
      pool.submit(new Runnable {
        override def run(): Unit = {
          client(i% 20) synchronized {
            client(i % 20).sayHello(name)
            latch.countDown()
          }
        }
      })
    }
    latch.await()
    println("took: " + (System.nanoTime() - t) / 1000000 + " ms")
    pool.shutdownNow()
  }
}

File: thrift/java/src/main/java/com/colobu/rpctest/AppClient.java
package com.colobu.rpctest;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;

public class AppClient {
    public static void main(String[] args) {
        try {
            TTransport[] transport = new TTransport[20];
            Greeter.Client[] client = new Greeter.Client[20];

            //warmup
            String name = "world";
            for (int i = 0; i < 20; i++) {
				transport[i] = new TSocket("localhost", 9090);
				transport[i].open();

				TProtocol protocol = new TBinaryProtocol(transport[i]);
                client[i] =  new Greeter.Client(protocol);
                client[i].sayHello(name);
            }

            boolean sync = true;
			if (args.length > 0) {
                sync = Boolean.parseBoolean(args[0]);
            }
            if (sync) { syncTest(client[0], name); }
            else { asyncTest(client, name); }

			for (int i = 0; i < 20; i++) {
				transport[i].close();
			}
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void syncTest(Greeter.Client client, String name) throws TException {
        long t = System.nanoTime();
        for (int i = 0; i < 10000; i++) {
            client.sayHello(name);
        }
        System.out.println("took: " + (System.nanoTime() - t) / 1000000 + " ms");
    }

    private static void asyncTest(final Greeter.Client[] client, final String name) throws InterruptedException {
        ExecutorService pool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
        final CountDownLatch latch = new CountDownLatch(10000);
        long t = System.nanoTime();
        for (int i = 0; i < 10000; i++) {
            final int j = i;
            pool.submit(new Runnable() {
                @Override
                public void run() {
                    synchronized (client[j % 20]) {
                        try {
                            client[j % 20].sayHello(name);
                        }
                        catch (TException e) {
                            e.printStackTrace();
                        }
                        finally {
                            latch.countDown();
                        }
                    }

                }
            });

        }

        latch.await();
        System.out.println("took: " + (System.nanoTime() - t) / 1000000 + " ms");
		pool.shutdownNow();
    }
}
